### PR TITLE
Ink 2 registration and login system

### PR DESCRIPTION
**Pull request by shroukhegazi**
- Created a User Registration system.
- Created User Login System.
- Sent Verification Mail.
- Write Test Cases to Test all Views.

### CODE REVIEW

**Review by MoazSamy — APPROVED** [review]
LGTM, I just think it'd be better to separate authentication urls/tests in a separate app 😄

File: config/settings/base.py [hunk 1]
```diff
@@ -219,6 +220,8 @@
     "DJANGO_EMAIL_BACKEND",
     default="django.core.mail.backends.smtp.EmailBackend",
 )
+# EMAIL_BACKEND = 'django.core.mail.backends.smtp.EmailBackend'
```
Comment (ahmedeltaweel) [inline config/settings/base.py:223]: this should have no effect on the current PR. please revert

File: config/settings/base.py [hunk 2]
```diff
@@ -325,3 +326,6 @@
 }
 # Your stuff...
 # ------------------------------------------------------------------------------
+
+EMAIL_HOST = "localhost"
```
Comment (ahmedeltaweel) [inline config/settings/base.py:330]: this exists in local.py. please remove.

File: config/urls.py [hunk 3]
```diff
@@ -16,6 +16,7 @@
     path(settings.ADMIN_URL, admin.site.urls),
     # User management
     path("users/", include("theinkspot.users.urls", namespace="users")),
```
Comment (ahmedeltaweel) [inline config/urls.py:18]: remove views. we only work with apis

File: requirements/base.txt [hunk 4]
```diff
@@ -22,3 +22,5 @@ djangorestframework==3.13.1  # https://github.com/encode/django-rest-framework
 django-cors-headers==3.11.0 # https://github.com/adamchainz/django-cors-headers
 # DRF-spectacular for api documentation
 drf-spectacular==0.21.2
+djangorestframework-simplejwt==5.1.0
+pytest-mock==3.8.2
```
Comment (ahmedeltaweel) [inline requirements/base.txt:26]: base is only for common packages

File: theinkspot/users/api/helpers.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import re
+
+
+# This function is checking the strgnth of a password
+# to contain a number and both uppercase and lowercase letters
+def check_password_strength(password):
+    if not (re.fullmatch(r"[A-Za-z0-9@#$%^&+=]{8,}", password)):
```
Comment (ahmedeltaweel) [inline theinkspot/users/api/helpers.py:7]:
```suggestion
    return not (re.fullmatch(r"[A-Za-z0-9@#$%^&+=]{8,}", password))
```

File: theinkspot/users/api/serializers.py [hunk 6]
```diff
@@ -1,14 +1,78 @@
...
+from .helpers import check_password_strength
+
 User = get_user_model()
 
 
 class UserSerializer(serializers.ModelSerializer):
     class Meta:
         model = User
-        fields = ["username", "name", "url"]
+        fields = "__all__"
```
Comment (ahmedeltaweel) [inline theinkspot/users/api/serializers.py:12]: Please be explicit as possible

Comment (ahmedeltaweel) [reply theinkspot/users/api/serializers.py:12]: one serializer for user is enough as model serializer

File: theinkspot/users/api/urls.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    router = DefaultRouter()
+else:
+    router = SimpleRouter()
+
+
+app_name = "api-users"
+urlpatterns = router.urls
+
+urlpatterns = [
+    path("register/", RegisterUsers.as_view(), name="register"),
```
Comment (ahmedeltaweel) [inline theinkspot/users/api/urls.py:18]: you can consider Router

File: theinkspot/users/api/views.py [hunk 8]
```diff
@@ -1,25 +1,78 @@
+# from email.message import _PayloadType
```
Comment (ahmedeltaweel) [inline theinkspot/users/api/views.py:1]: commented imports should be removed

File: theinkspot/users/tests/test_views.py [hunk 9]
```diff
@@ -7,13 +7,18 @@
 from django.http import HttpRequest, HttpResponseRedirect
 from django.test import RequestFactory
 from django.urls import reverse
+from rest_framework.test import APIClient
```
Comment (ahmedeltaweel) [inline theinkspot/users/tests/test_views.py:10]: revert here

**Review by ahmedeltaweel — CHANGES_REQUESTED** [review]

**Review by ahmedeltaweel — APPROVED** [review]

**Pull request merged**